crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1654,48 +1654,56 @@ impl<'db> TypeInferenceBuilder<'db> {
     fn infer_fstring_expression(&mut self, fstring: &ast::ExprFString) -> Type<'db> {
         let ast::ExprFString { range: _, value } = fstring;
 
+        let mut done = false;
+        let mut has_expression = false;
+        let mut concatenated = String::new();
         for part in value {
             match part {
-                ast::FStringPart::Literal(_) => {
-                    // TODO string literal type
+                ast::FStringPart::Literal(literal) => {
+                    concatenated.push_str(&literal.value);
                 }
                 ast::FStringPart::FString(fstring) => {
-                    let ast::FString {
-                        range: _,
-                        elements,
-                        flags: _,
-                    } = fstring;
-                    for element in elements {
-                        self.infer_fstring_element(element);
+                    for element in &fstring.elements {
+                        match element {
+                            ast::FStringElement::Expression(expression) => {
+                                // TODO: handle the format-spec in expression
```

Comment:
This might be quite complex to do fully, and may not be a priority for quite some time, if ever. But I don't want us to give _incorrect_ results in the meantime. It seems like it should be quite easy to detect whether there is a format string at all, and if so, just fall back to `LiteralString` instead of a specific `StringLiteral` type. That seems worth doing now?

We should still have a TODO for handling `__format__` and format specifiers (at least for some types); I'd like that TODO to specifically mention `__format__`.

Reply:
That's done - I realised however that there's more work than I expected there

- Conversion flags should be handled before calling format
- Format logic could be re-coded specifically for builtins
- From [this](https://docs.python.org/3/library/functions.html#format), we should keep track of non-empty format specifiers reaching the `object` implementation
- ...
> Changed in version 3.4: object().__format__(format_spec) raises [TypeError](https://docs.python.org/3/library/exceptions.html#TypeError) if format_spec is not an empty string.

Reply:
That's done - I realised however that there's more work than I expected there

- Conversion flags should be handled before calling format
- Format logic could be re-coded specifically for builtins
- From [this](https://docs.python.org/3/library/functions.html#format), we should keep track of non-empty format specifiers reaching the `object` implementation
> Changed in version 3.4: object().__format__(format_spec) raises [TypeError](https://docs.python.org/3/library/exceptions.html#TypeError) if format_spec is not an empty string.